### PR TITLE
Add vsvimatomicinsert setting to fix Intellisense repeations

### PR DESCRIPTION
**Pull request by fredizzimo**
Hi Jared,

I have added a non-standard vim setting vsvimatomicinsert (ains), which when enabled fixes most of the issues with repeating commands that uses Visual Assist code completion. The primary issue that it fixes is #664, but probably not completely.

When the vsatomicinsert option is enabled, then all movements in insert mode is tracked. All absolute movements (by the mouse or Visual Assist for example), are converted into relative ones. So you can repeat the action in a different location.

I also added a couple of unit test for the normal insert mode, since no test broke, when I changed the code, when it clearly should have broken something.

I have not tested this very extensively yet, since I just finished it. But I'm creating this pull request so that you can look at it and comment. Unfortunately, I'm not going to write much c++ at work during the following weeks, so I won't be able to test it properly during that time either.

However most basic tests seems to work, even multi-line code snippet repetitions, that opens a popup to ask for some information. They can't be repeated into a block with a different indentation though, and I don't have any idea how to fix that, I also think I saw some existing bug about that. 

Another thing that breaks is, if you have the "Insert _ after m and shift" option enabled. It works by highlighting the the _ character, while shift is pressed, so that it will be overwritten if you press _. When you then release shift, it stops the highlight and moves the cursor forward. This could probably be detected and fixed, but I think much better, is to just disable that option. At least for me, I don't see how it's very useful, 

Do you have some comments about the naming of the option? And should we have separate options for tracking direct cursor movements(with mouse or automatic), and movements done by the arrow keys?

I have no previous F# experience, and almost no C# experience either, so please tell me if you see something that could be done in a better way.

Edit:
I forgot to mention, that this doesn't fix macros containing code completion. That has to be fixed in a different way.

Best Regards,
Fred Sundvik

### CODE REVIEW

**Comment by academician** [comment]
Hey, Fred and Jared - if this is an actual fix to some of the Visual Assist integration problems I've been having, I'd love to see it merged. Any plans to look at it?

**Comment by fredizzimo** [comment]
Hi academician,

I have been running with this enabled since I made the merge request, and everything seems to be working great for me. I haven't actually encountered any problems at all for my use cases, the dot repetition seems to work, even for multiline completions.

However, it has to be noted that the behavior is different from normal Vim, so therefore I added a new option ("vsvimatomicinsert"). Without that option everything should behave as it did before.

It also doesn't fix all the autocompletion problems with recorded macros. But those should be possible to fix in a similar way.

From my point of view, this change can be merged, if Jared agrees on the naming of the option, and the actual implementation. It would of course be great if someone else could test it first as well.

File: Src/VimCore/VimSettings.fs [hunk 1]
```diff
@@ -149,6 +149,7 @@ type internal GlobalSettings() =
 
     static let GlobalSettingInfoList = 
         [|
+            (AtomicInsertName, "ains", SettingValue.Toggle false)
```
Comment (jaredpar) [inline Src/VimCore/VimSettings.fs]: Can we use the pattern `(AtomicInsertName, AtomicInsertName, ...)` here?  In general for non-standard options, like `CaretOpacityName`, I don't provide abbreviations.  The reasoning is to try and keep non-standard options as visible as possible.

**Comment by jaredpar** [comment]
@fredizzimo sorry.  I completely lost track of this PR after you initially submitted it.  I just looked it over and think it's mostly good to go.  Provided a minor comment about some naming issues but otherwise the change looks great!

**Comment by fredizzimo** [comment]
Hi Jared,

Sorry for the delay, I had some problems with the computer at home. But I have now updated the pull request with the short name of the option removed.

**Comment by jaredpar** [comment]
Thanks for the contribution!

**Pull request merged**